### PR TITLE
[SPARK-22959][PYTHON] Configuration to select the modules for daemon and worker in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

We are now forced to use `pyspark/daemon.py` and `pyspark/worker.py` in PySpark.

This doesn't allow a custom modification for it (well, maybe we can still do this in a super hacky way though, for example, setting Python executable that has the custom modification). Because of this, for example, it's sometimes hard to debug what happens inside Python worker processes.

This is actually related with [SPARK-7721](https://issues.apache.org/jira/browse/SPARK-7721) too as somehow Coverage is unable to detect the coverage from `os.fork`. If we have some custom fixes to force the coverage, it works fine.

This is also related with [SPARK-20368](https://issues.apache.org/jira/browse/SPARK-20368). This JIRA describes Sentry support which (roughly) needs some changes within worker side. 

With this configuration advanced users will be able to do a lot of pluggable workarounds and we can meet such potential needs in the future.

As an example, let's say if I configure the module `coverage_daemon` and had `coverage_daemon.py` in the python path:

```python
import os

from pyspark import daemon


if "COVERAGE_PROCESS_START" in os.environ:
    from pyspark.worker import main

    def _cov_wrapped(*args, **kwargs):
        import coverage
        cov = coverage.coverage(
            config_file=os.environ["COVERAGE_PROCESS_START"])
        cov.start()
        try:
            main(*args, **kwargs)
        finally:
            cov.stop()
            cov.save()
    daemon.worker_main = _cov_wrapped


if __name__ == '__main__':
    daemon.manager()
```

I can track the coverages in worker side too.

More importantly, we can leave the main code intact but allow some workarounds.

## How was this patch tested?

Manually tested.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@holdenk, @rxin, @joshrosen and @ueshin,  as you all might already know, I am working on Python coverage. Based on the top of this PR, I think we can leave the main codes intact while we properly track the coverage within worker processes. I believe this also partly covers SPARK-20368 too.

What do you guys think about this configuration?

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: generally, I thought we use the name "command" as what we call the thing to execute

**Review by felixcheung — COMMENTED** [review]
generally LGTM

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Ah, yup that's true in general. But please let me stick to "module" here as that's what we execute (`python -m`) describes:

```
python --help
...
-m mod : run library module as a script (terminates option list)
...
```

**Comment by HyukjinKwon** [comment]
Hey @rxin, I think I need your sign-off too as it's related with SPARK-7721.

**Comment by ueshin** [comment]
The changes LGTM.
Btw, what if we miss the module in python path? Can we see that the error is caused by the missing module from the exception message?

**Comment by HyukjinKwon** [comment]
I manually tested after setting `spark.python.daemon.module` to `nonexistantmodule`. It shows the error message like this:

```python
>>> spark.range(1).rdd.map(lambda x: x).collect()
```

```
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/rdd.py", line 824, in collect
    port = self.ctx._jvm.PythonRDD.collectAndServe(self._jrdd.rdd())
  File "/.../spark/python/lib/py4j-0.10.6-src.zip/py4j/java_gateway.py", line 1160, in __call__
  File "/.../spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/.../spark/python/lib/py4j-0.10.6-src.zip/py4j/protocol.py", line 320, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 0.0 failed 1 times, most recent failure: Lost task 1.0 in stage 0.0 (TID 1, localhost, executor driver): org.apache.spark.SparkException:
Error from python worker:
  /usr/bin/python: No module named nonexistantmodule
PYTHONPATH was:
  /.../spark/python/lib/pyspark.zip:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/assembly/target/scala-2.11/jars/spark-core_2.11-2.3.0-SNAPSHOT.jar:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/python/:
java.io.EOFException
    ...

Driver stacktrace:
    ...
Caused by: org.apache.spark.SparkException:
Error from python worker:
  /usr/bin/python: No module named nonexistantmodule
PYTHONPATH was:
  /.../spark/python/lib/pyspark.zip:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/assembly/target/scala-2.11/jars/spark-core_2.11-2.3.0-SNAPSHOT.jar:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/python/:
java.io.EOFException
    ...
    ... 1 more

18/01/08 15:54:06 WARN TaskSetManager: Lost task 6.0 in stage 0.0 (TID 6, localhost, executor driver): TaskKilled (Stage cancelled)
...
```

**Comment by ueshin** [comment]
Looks good. Let's wait for @rxin's response.

**Comment by HyukjinKwon** [comment]
Yup, thanks for all review @felixcheung and @ueshin BTW

**Comment by HyukjinKwon** [comment]
@rxin or @joshrosen could you guys take a quick look and see if it makes sense?

**Comment by holdenk** [comment]
So I think this could be the basis for solving a lot of related problems and I like the minimally invasive approach to it. I think the error message for setting it to a bad module rather than a nonexistent module is probably going to be very confusing. I think it would be good to make it clear that this is advanced setting we don't expect most users to modify directly.

**Comment by felixcheung** [comment]
+1... this is "undocumented" conf, sooo it's an expert one :)

**Comment by HyukjinKwon** [comment]
Yup, will write up some more warnings that says like it's expert only configuration, experimental and rather an internal configuration. Also, I will note that we should be super careful. Will update tonight (KST) :).

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Just double checked it shows the log only when the configuration is explicitly set:

```
18/01/10 21:23:24 INFO PythonWorkerFactory: Python daemon module in PySpark is set to [pyspark.daemon] in
 'spark.python.daemon.module', using this to start the daemon up. Note that this configuration only has an 
effect when 'spark.python.use.daemon' is enabled and the platform is not Windows.
```

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
Comment (viirya) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Do we need to restrict the module's package to only allow something like `pyspark.*`?

**Review by BryanCutler — APPROVED** [review]
LGTM, I think there have been many times that this would have been incredibly useful for me, thanks!

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Hm, actually we could check like .. if it's empty string too. I wrote "shouldn't be used before knowing what it means exactly." above. So, I think it's fine.

**Comment by HyukjinKwon** [comment]
Will merge this one if there isn't any objection. I believe this doesn't affect the existing code path anyway ..

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
